File: package/go/src/hit_and_blow.go
package src

// HitAndBlowResult represents the result of Hit & Blow game
type HitAndBlowResult struct {
	Hits  int
	Blows int
}

// HitAndBlow performs Hit & Blow game judgment
func HitAndBlow(answer, guess []int) HitAndBlowResult {
	// Count hits (same position and same value)
	hits := 0
	for i := 0; i < len(answer); i++ {
		if answer[i] == guess[i] {
			hits++
		}
	}

	// Count frequency of each number in both arrays
	answerCounter := make(map[int]int)
	guessCounter := make(map[int]int)

	for _, num := range answer {
		answerCounter[num]++
	}

	for _, num := range guess {
		guessCounter[num]++
	}

	// Count total matches (regardless of position)
	totalMatches := 0
	for num, answerCount := range answerCounter {
		guessCount := guessCounter[num]
		if guessCount < answerCount {
			totalMatches += guessCount
		} else {
			totalMatches += answerCount
		}
	}

	// Blows = total matches - hits
	blows := totalMatches - hits

	return HitAndBlowResult{
		Hits:  hits,
		Blows: blows,
	}
}

File: package/rust/src/hit_and_blow.rs
pub fn hit_and_blow(answer: &[i32], guess: &[i32]) -> (i32, i32) {
    let mut hits = 0;
    let mut answer_counts = std::collections::HashMap::new();
    let mut guess_counts = std::collections::HashMap::new();

    for i in 0..answer.len().min(guess.len()) {
        if answer[i] == guess[i] {
            hits += 1;
        } else {
            *answer_counts.entry(answer[i]).or_insert(0) += 1;
            *guess_counts.entry(guess[i]).or_insert(0) += 1;
        }
    }

    let mut common_count = 0;
    for (&value, &answer_count) in &answer_counts {
        if let Some(&guess_count) = guess_counts.get(&value) {
            common_count += answer_count.min(guess_count);
        }
    }

    let blows = common_count;

    (hits, blows)
}
